pkg/domain/stateDiffComponent.go

```diff
@@ -0,0 +1,187 @@
+package domain
+
+import (
+	"fmt"
+	"github.com/go-logr/logr"
+	"golang.org/x/exp/maps"
+
+	"github.com/cloudogu/cesapp-lib/core"
+	"github.com/cloudogu/k8s-blueprint-operator/pkg/domain/ecosystem"
+)
+
+// ComponentDiffs contains the differences for all expected Components to the current ecosystem.ComponentInstallations.
+type ComponentDiffs []ComponentDiff
+
+// Statistics aggregates various figures about the required actions of the ComponentDiffs.
+func (cd ComponentDiffs) Statistics() (toInstall int, toUpgrade int, toUninstall int, other int) {
+	for _, componentDiff := range cd {
+		switch componentDiff.NeededAction {
+		case ActionInstall:
+			toInstall += 1
+		case ActionUpgrade:
+			toUpgrade += 1
+		case ActionUninstall:
+			toUninstall += 1
+		default:
+			other += 1
+		}
+	}
+	return
+}
+
+// ComponentDiff represents the Diff for a single expected Component to the current ecosystem.ComponentInstallation.
+type ComponentDiff struct {
+	// Name contains the component's name.
+	Name string
+	// Actual contains that state of a component how it is currently found in the system.
+	Actual ComponentDiffState
+	// Expected contains that desired state of a component how it is supposed to be.
+	Expected ComponentDiffState
+	// NeededAction hints how the component should be handled by the application change automaton in order to reconcile
+	// differences between Actual and Expected in the current system.
+	NeededAction Action
+}
+
+// ComponentDiffState contains all fields to make a diff for components in respect to another ComponentDiffState.
+type ComponentDiffState struct {
+	// DistributionNamespace is part of the address under which the component will be obtained. This namespace does NOT
+	// to be confused with the K8s cluster namespace.
+	DistributionNamespace string
```

Comment:
Namespace switches are not considered for the diff yet, correct?
This is okay, since we can add those later.

Reply:
That's what I thought as well